src/linter/block.go

```diff
@@ -64,10 +64,13 @@ type BlockWalker struct {
 	// state
 	statements map[node.Node]struct{}
 
-	// whether a function has a return without explit expression.
+	// whether a function has a return without explicit expression.
 	// Required to make a decision in void vs null type selection,
 	// since "return" is the same as "return null".
 	bareReturn bool
+	// stores return nodes that have no return value.
+	bareReturns []*stmt.Return
```

Comment:
It's more reliable to check whether a function return type `!typ.Is("void")` than depend on the `bareReturn` flag in my opinion. This will also make it possible to generate the report at the point where you're handling that return (so the slice is unnecessary in that case).

One of the corner case is this:
```php
/**
 * @return null
 */
function f() {
  return;
}
```
The `returnsValue` would be false, so you can't warn on `return` in this case even though the function is intended to be a non-void function (it returns null).

P.S. if we infer `f` function return type as void, that's a bug (it should be `null`).